### DISCUSSION TITLE
Splitting image vertically using a specific color?

### DISCUSSION THREAD

**Discussion by manpaint** [ImageMagick/ImageMagick#4829] [Help] [answered]
**EDIT: See my last post for the complete answers.**

After searching the web and the documentation, I could not find how to do this. I have some sort of vertical table with irregular heights. The only way I can see to split it would be to detect the color rbg(0,255,33) (very bright green) but I am unsure how to this.

Here's an example of what I am trying to achieve:

![Example](https://i.imgur.com/aAagrkj.png)

I suspect there may be a way to do it with -crop but I can only a way to specify a specific amount of pixels, which is not what I need.

I would appreciate if anyone could give me some pointers on how to achieve this.

**ImageMagick ver:** ImageMagick-7.0.6-Q16
**Operating System:** Windows 7 (can use Windows 10 if required)

**Comment by fmw42** [comment]
Please post the original image by itself.  I suggest you threshold on the green color. Then use connected-components to find the bounding boxes. From those you can get the vertical coordinates or just use the bounding boxes of the non-green to crop your image. See https://imagemagick.org/script/connected-components.php

**Reply by manpaint** [reply]
Here's the original image:

![start.png](https://i.imgur.com/XuYgyDD.png)

I tried to use connected components, but it does not appear to be able to detect the green lines (I am probably doing something wrong).

``` 
"C:\Program Files\ImageMagick-7.0.6-Q16\magick.exe" start.png -connected-components 1 -auto-level -depth 1 objects.png
```

![object.png](https://i.imgur.com/6VFBTca.png)

After another dive in the documentation, the closest I got to my objective was following command:

``` 
"C:\Program Files\ImageMagick-7.0.6-Q16\magick.exe" convert start.png -trim "rgb(0,255,33)" +repage 0-%03d.png
```
Result:

![0-000.png](https://i.imgur.com/7S92lkf.png)

(The very top and bottom green lines are now gone)

It's not quite what I want yet but it's the closest I got. Do you have any idea how I should proceed forward?

**Comment by fmw42** [comment]
Here is how to do that in Unix.  Sorry, I do not know Windows scripting.

Input:

![68747470733a2f2f692e696d6775722e636f6d2f587559677944442e706e67](https://user-images.githubusercontent.com/13458218/153663270-db7a9ed2-c788-4178-a52c-71b775b6c5dd.png)

```
bboxArr=(`magick 68747470733a2f2f692e696d6775722e636f6d2f587559677944442e706e67.png \
-color-threshold "srgb(0,240,0)-srgb(0,255,50)" -type bilevel \
-define connected-components:verbose=true \
-define connected-components:mean-color=true \
-connected-components 8 null: | grep "gray(0)" | awk '{print $2}'`)
num=${#bboxArr[*]}
for ((i=0; i<num; i++)); do
bbox="${bboxArr[$i]}"
magick 68747470733a2f2f692e696d6775722e636f6d2f587559677944442e706e67.png -crop $bbox +repage result_$i.png
done

```

Results:

![result_0](https://user-images.githubusercontent.com/13458218/153663320-8c8ada3c-9005-41af-9620-dcdd58aa3073.png)

![result_1](https://user-images.githubusercontent.com/13458218/153663330-cc0c5334-0540-4213-af79-3db9d85aa361.png)

![result_2](https://user-images.githubusercontent.com/13458218/153663335-db022238-19b5-4305-846f-5025311d08a8.png)

**Comment by snibgo** [comment]
fmw42's script, modified for Windows BAT, is:
```
for /F "usebackq" %%L in (`magick greenLines3.png ^
-color-threshold "srgb(0,240,0)-srgb(0,255,50)" -type bilevel ^
-define "connected-components:verbose=true" ^
-define "connected-components:mean-color=true" ^
-connected-components 8 x1.png ^| grep "gray(0)" ^| gawk '{print $2}'`) do magick ^
greenLines3.png ^
-crop %%L +repage result_%%L.png
```
My computer has "gawk" instead of "awk".

The output filenames include the crop parameters, for example "result_139x106+0+1.png".

EDIT: removed superfluous "%IMG7".

**Reply by fmw42** [reply]
I suspect that you would need to remove "%IMG7%" since I think snibgo uses that so that he can have two versions of imagemagick on his system. So just "magick" for one version of ImageMagick 7 and no ImageMagick 6 on the system.

**Reply by manpaint** [reply]
I tried to execute the following the .bat script as instructed (with only filenames modified):

```
@echo off
"C:\Program Files\ImageMagick-7.0.6-Q16\magick.exe"
for /F "usebackq" %%L in (`magick start.png ^
-color-threshold "srgb(0,240,0)-srgb(0,255,50)" -type bilevel ^
-define "connected-components:verbose=true" ^
-define "connected-components:mean-color=true" ^
-connected-components 8 x1.png ^| grep "gray(0)" ^| awk '{print $2}'`) do magick ^
start.png ^
-crop %%L +repage result_%%L.png
pause
```

It returned this error:

```
Error: Invalid argument or not enough arguments

Usage: magick.exe [ {option} | {image} ... ] {output_image}
       magick.exe [ {option} | {image} ... ] -script {filename} [ {script_args}
...]
       magick.exe -help | -version | -usage | -list {option}

'gawk' is not recognized as a command or executable.
```

The error seems to indicate that the gawk part is the problem. I tried to install GnuWin32 ( https://sourceforge.net/projects/gnuwin32/ ) to remedy this but it did not do anything.

The awk variant do not work either. Do you guys have any idea of the root issue?

**Reply by fmw42** [reply]
I do not know how to script on Windows. But I do not understand why you have 

> "C:\Program Files\ImageMagick-7.0.6-Q16\magick.exe"

that line in your listing. Try removing it. Or wait for snibgo to help you.

**Comment by snibgo** [comment]
Thanks for the correction, fmw42.

**Comment by snibgo** [comment]
@manpaint:

When getting a script to work, "@echo off" just makes the job harder, because you can't see which command causes which problems.

Your BAT script contains this line:
```
"C:\Program Files\ImageMagick-7.0.6-Q16\magick.exe"
```
So you run "magick" with no arguments at all. Why? This gives the error message you show: "Error: Invalid argument or not enough arguments".

If you don't have awk or gawk or similar, the job can still be done, but it is slightly more complex.

**Reply by manpaint** [reply]
I altered the .bat files following the recommendations. For some reason, I though that the "magick" command could not exist in a vacuum and that binary file had to invoked. Still, CMD do not detect gawk or awk, so I tried the following:

```
for /F "usebackq" %%L in (`magick start.png ^
-color-threshold "srgb(0,240,0)-srgb(0,255,50)" -type bilevel ^
-define "connected-components:verbose=true" ^
-define "connected-components:mean-color=true" ^
-connected-components 8 x1.png ^| grep "gray(0)" ^| "C:\Users\Pc\Desktop\autre dossiers\Tools\Gawk\bin\gawk.exe" gawk '{print $2}'`) do magick ^
start.png ^
-crop %%L +repage result_%%L.png
pause
```

It now returns this error:
```
C:\Users\Pc\Desktop\image magick test\2>for /F "usebackq" %L in (`magick start.p
ng -color-threshold "srgb(0,240,0)-srgb(0,255,50)" -type bilevel -define "connec
ted-components:verbose=true" -define "connected-components:mean-color=true" -con
nected-components 8 x1.png | grep "gray(0)" | "C:\Users\Pc\Desktop\autre dossier
s\Tools\Gawk\bin\gawk.exe" gawk '{print $2}'`) do magick start.png -crop %L +rep
age result_%L.png
magick: unrecognized option `-color-threshold' at CLI arg 2 @ fatal/magick-cli.c
/ProcessCommandOptigoanwsk/:4 25c.m
. line:1: fatal: cannot open file `'{print' for reading (No such file or directo
ry)
```

I tried setting environment variables and reinstalling Gawk, but it did not change anything. I would like to see what the alternative to using Gack is as it seems that it is the only option remaining.

**Comment by snibgo** [comment]
"-color-threshold" was introduced in IM v7.0.9-27 (see https://www.imagemagick.org/discourse-server/viewtopic.php?f=4&t=37620 ). You use v7.0.6, so you need to upgrade your IM for that feature.

**Reply by manpaint** [reply]
After upgrading Image Magick, I now get the following error (using the same BAT script as earlier):

```
C:\Users\Pc\Desktop\image magick test\2>for /F "usebackq" %L in (`magick start.p
ng -color-threshold "srgb(0,240,0)-srgb(0,255,50)" -type bilevel -define "connec
ted-components:verbose=true" -define "connected-components:mean-color=true" -con
nected-components 8 x1.png | grep "gray(0)" | "C:\Users\Pc\Desktop\autre dossier
s\Tools\Gawk\bin\gawk.exe" gawk '{print $2}'`) do magick start.png -crop %L +rep
age result_%L.png
gawk: cmd. line:1: fatal: cannot open file `'{print' for reading (No such file o
r directory)
grep: write error: Illegal seek
```

It seems to think that '{print' is a filename for reasons that elude me.

**Reply by fmw42** [reply]
You have gawk twice.  Either remove `"C:\Users\Pc\Desktop\autre dossiers\Tools\Gawk\bin\gawk.exe"`  or remove `gawk`

**Reply by manpaint** [reply]
It appears that it returns the same error even when gawk is removed.

Script:
```
for /F "usebackq" %%L in (`magick start.png ^
-color-threshold "srgb(0,240,0)-srgb(0,255,50)" -type bilevel ^
-define "connected-components:verbose=true" ^
-define "connected-components:mean-color=true" ^
-connected-components 8 x1.png ^| grep "gray(0)" ^| "C:\Users\Pc\Desktop\autre dossiers\Tools\Gawk\bin\gawk.exe" '{print $2}'`) do magick ^
start.png ^
-crop %%L +repage result_%%L.png
pause
```

Error:

```
C:\Users\Pc\Desktop\image magick test\2>for /F "usebackq" %L in (`magick start.p
ng -color-threshold "srgb(0,240,0)-srgb(0,255,50)" -type bilevel -define "connec
ted-components:verbose=true" -define "connected-components:mean-color=true" -con
nected-components 8 x1.png | grep "gray(0)" | "C:\Users\Pc\Desktop\autre dossier
s\Tools\Gawk\bin\gawk.exe" '{print $2}'`) do magick start.png -crop %L +repage r
esult_%L.png
gawk: '{print
gawk: ^ invalid char ''' in expression
grep: write error: Illegal seek
```

**Reply by fmw42** [reply]
You should test your gawk in stand alone mode to be sure you are calling it correctly and it works.

**Comment by manpaint** [comment, marked as answer]
When I looked at example script for Gawk, I was able to figure out why it did not work. {print $2} needed to be surrounded by " and not '. 
**Therefore, the final answer is the following:**

On a Windows 7 machine with ImageMagick 7.1.0-24 and Gawk, it is possible to realize what is described in the first post using the following BAT script (this is for CMD, NOT Powershell):

```
for /F "usebackq" %%L in (`magick start.png ^
-color-threshold "srgb(0,240,0)-srgb(0,255,50)" -type bilevel ^
-define "connected-components:verbose=true" ^
-define "connected-components:mean-color=true" ^
-connected-components 8 x1.png ^| grep "gray(0)" ^| "C:\Users\Pc\Desktop\autre dossiers\Tools\Gawk\bin\gawk.exe" "{print $2}"`) do magick ^
start.png ^
-crop %%L +repage 0zzzzzz_%%L.png
pause
```

I thank everyone who helped me with this problem. I hope you all have a good day.

**Reply by snibgo** [reply]
On my Windows 8.1 laptop, either single or double quotes works fine around "{print $2}". Windows 7 may be more restrictive.
